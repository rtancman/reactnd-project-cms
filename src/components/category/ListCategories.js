import React, { Component } from 'react';
import PropTypes from 'prop-types'
import { connect } from 'react-redux';
import CircularProgress from '@material-ui/core/CircularProgress';
import { categoriesFetchData } from './actions';
import { Link } from 'react-router-dom';
import { List } from 'components/layout/List'

class ListCategories extends Component {
  static propTypes = {
    fetchData: PropTypes.func.isRequired,
    items: PropTypes.array.isRequired,
    isFetching: PropTypes.bool.isRequired,
    didInvalidate: PropTypes.bool.isRequired
  }

  componentDidMount() {
    this.props.fetchData()
  }

  render() {
    let content = ''
    const { items, isFetching, didInvalidate } = this.props

    if (didInvalidate) {
      content = (<p>Sorry! There was an error loading the items</p>)
    } else if (isFetching) {
      content = (
        <div>
          <CircularProgress thickness={7} />
          <p>Loadingâ€¦</p>
        </div>
      )
    } else if ( items.length > 0 ) {
      content = (
        <List title='Categories' items={items} />
      )
    }

    return (
      <div className='ListCategories'>
        { content }
      </div>
    );
  }
}

const mapStateToProps = ({ categories }) => {
  return {
    items: categories.items,
    didInvalidate: categories.didInvalidate,
    isFetching: categories.isFetching
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    fetchData: () => dispatch(categoriesFetchData())
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(ListCategories);
