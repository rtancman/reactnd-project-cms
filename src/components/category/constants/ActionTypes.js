export const REQUEST_CATEGORIES = 'REQUEST_CATEGORIES'
export const RECEIVE_CATEGORIES = 'RECEIVE_CATEGORIES'
export const INVALIDATE_CATEGORIES = 'INVALIDATE_CATEGORIES'

export const REQUEST_CATEGORY_POSTS = 'REQUEST_CATEGORY_POSTS'
export const RECEIVE_CATEGORY_POSTS = 'RECEIVE_CATEGORY_POSTS'
export const INVALIDATE_CATEGORY_POSTS = 'INVALIDATE_CATEGORY_POSTS'
