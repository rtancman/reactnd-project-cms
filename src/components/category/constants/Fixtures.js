export const categoriesMock = {
  categories: [{
    "name": "react",
    "path": "react"
    },
    {
      "name": "redux",
      "path": "redux"
    },
    {
      "name": "udacity",
      "path": "udacity"
    }
  ]
}

export const categoryPostsMock = [
  {
      "id": "8xf0y6ziyjabvozdd253nd",
      "timestamp": 1467166872634,
      "title": "Udacity is the best place to learn React",
      "body": "Everyone says so after all.",
      "author": "thingtwo",
      "category": "react",
      "voteScore": 6,
      "deleted": false,
      "commentCount": 2
  }
]