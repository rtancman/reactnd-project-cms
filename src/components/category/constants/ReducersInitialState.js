export const initialCategoriesState = {
  isFetching: false,
  didInvalidate: false,
  items: [],
}