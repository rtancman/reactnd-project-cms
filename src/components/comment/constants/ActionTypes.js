export const REQUEST_CREATE_COMMENT = 'REQUEST_CREATE_COMMENT'
export const INVALIDATE_CREATE_COMMENT = 'REQUEST_CREATE_COMMENT'
export const COMMENT_HAS_BEEN_CREATED = 'COMMENT_HAS_BEEN_CREATED'