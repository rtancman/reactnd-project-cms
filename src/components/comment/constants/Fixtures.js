export const commentMock = {
  "id": "AAAAb033-81ca-4bea-8f51-237ac1694d70",
  "timestamp": 1536967930576,
  "body": "Redux is amazing!\nRedux is amazing!\nRedux is amazing!\nRedux is amazing!\nRedux is amazing!\nRedux is amazing!\nRedux is amazing!Redux is amazing!\n\n:)",
  "author": "Raffael Tancman",
  "parentId": "e91ab033-81ca-4bea-8f51-237ac1694d70",
  "voteScore": 1,
  "deleted": false,
  "parentDeleted": false
}