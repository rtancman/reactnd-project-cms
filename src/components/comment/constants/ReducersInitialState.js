export const initialCreateCommentState = {
  isFetching: false,
  didInvalidate: false,
  created: false
}