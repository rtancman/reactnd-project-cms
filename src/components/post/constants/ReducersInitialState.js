export const initialListPostState = {
  isFetching: false,
  didInvalidate: false,
  items: [],
}

const post = {
  id: '',
  timestamp: 0,
  title: '',
  body: '',
  author: '',
  category: '',
  voteScore: 0,
  deleted: false,
  commentCount: 0
}

export const initialPostState = {
  isFetching: false,
  didInvalidate: false,
  content: {},
  comment: {}
}

export const initialCreatePostState = {
  isFetching: false,
  didInvalidate: false,
  created: false
}

export const initialRemovePostState = {
  isFetching: false,
  didInvalidate: false,
  removed: false,
  id: '',
}