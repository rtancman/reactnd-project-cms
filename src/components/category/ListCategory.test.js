import React from 'react'
import { shallow, mount } from 'enzyme'
import configureStore from 'redux-mock-store'
import fetchMock from 'fetch-mock'
import { MemoryRouter } from 'react-router'
import thunk from 'redux-thunk'
import ListCategories from './ListCategories'
import { initialCategoriesState } from './constants/ReducersInitialState'
import { categoriesMock } from './constants/Fixtures'
import { listCategoriesUrl } from '../../api/cms'
import { Provider } from 'react-redux'


describe('ListCategory Component', () => {
  const middlewares = [thunk]
  const mockStore = configureStore(middlewares)
  let store
  
  describe('when component renders', () => {
    beforeEach(() => {
      store = mockStore({categories: {...initialCategoriesState, items: categoriesMock.categories} })
      fetchMock.getOnce(listCategoriesUrl, { body: {categories: categoriesMock.categories} })
    })

    afterEach(() => {
      fetchMock.reset()
      fetchMock.restore()
    })
    
    it('without crashing', () => {
      const props = {
        store: store,
      }
      let wrapper = shallow(<ListCategories {...props} />)

      expect(wrapper).toMatchSnapshot()
    })

    it('when request categories displays the CircularProgress', () => {
      store = mockStore({categories: {...initialCategoriesState, items: []} })
      const root = mount(
        <Provider store={store}>
          <MemoryRouter>
            <ListCategories />
          </MemoryRouter>
        </Provider>
      )

      let wrapper = root.find('ListCategories')
      expect(wrapper.length).toBe(1)
      wrapper.instance().setState({isFetching: true})
      root.update()
      wrapper = root.find('ListCategories')
      expect(wrapper.find('CircularProgress').length).toBe(1)
      expect(wrapper.props().items.length).toBe(0)
      expect(wrapper.find('List').length).toBe(0)
    })
  
    it('when receive categories displays the List', () => {
      const root = mount(
        <Provider store={store}>
          <MemoryRouter>
            <ListCategories />
          </MemoryRouter>
        </Provider>
      )

      let wrapper = root.find('ListCategories')
      expect(wrapper.length).toBe(1)
      wrapper.instance().setState({isFetching: false})
      root.update()
      wrapper = root.find('ListCategories')
      expect(wrapper.find('CircularProgress').length).toBe(0)
      expect(wrapper.find('List').length).toBe(1)
    })
  
    it('when does not receive categories show the error message', () => {
      const root = mount(
        <Provider store={store}>
          <MemoryRouter>
            <ListCategories />
          </MemoryRouter>
        </Provider>
      )

      let wrapper = root.find('ListCategories')
      expect(wrapper.length).toBe(1)
      wrapper.instance().setState({
        isFetching: false,
        didInvalidate: true,
      })
      root.update()
      wrapper = root.find('ListCategories')
      expect(wrapper.find('p').length).toBe(1)
      expect(wrapper.find('p').text()).toBe('Sorry! There was an error loading the items')
      expect(wrapper.find('CircularProgress').length).toBe(0)
    })
  })
})