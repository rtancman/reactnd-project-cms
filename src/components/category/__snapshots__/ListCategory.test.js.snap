// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListCategory Component when component renders without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(ListCategories)
    store={
        Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "dispatch": [Function],
      "items": Array [
        Object {
          "name": "react",
          "path": "react",
        },
        Object {
          "name": "redux",
          "path": "redux",
        },
        Object {
          "name": "udacity",
          "path": "udacity",
        },
      ],
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "storeSubscription": Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "dispatch": [Function],
        "items": Array [
          Object {
            "name": "react",
            "path": "react",
          },
          Object {
            "name": "redux",
            "path": "redux",
          },
          Object {
            "name": "udacity",
            "path": "udacity",
          },
        ],
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "storeSubscription": Subscription {
          "listeners": Object {
            "clear": [Function],
            "get": [Function],
            "notify": [Function],
            "subscribe": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "unsubscribe": [Function],
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
